integrations/access/accessrequest/app.go

```diff
@@ -146,7 +149,7 @@ func (a *App) run(ctx context.Context) error {
 	}
 
 	if err := a.initAccessMonitoringRulesCache(ctx); err != nil {
-		return trace.Wrap(err)
+		log.WithError(err).Errorf("initialising Access Monitoring Rule cache")
 	}
```

Comment:
Instead of blindly asking for AccessMonitoringRules, we can check the watcher OpInit event and get its allowed resources.
If it returns AccessMonitoringRule kind then we are safe to call the List API.
It should be something like this:
https://github.com/gravitational/teleport/blob/b151260398da7a6218a286ee8a4bf4780d99ba9a/lib/cache/cache.go#L1246-L1256
(I don't think you need the `else` block)

Reply:
Something like this?

Reply:
Yeah
I don't like the channel thing, seems to add unnecessary complexity.
But I'm failing to find a better solution.

Changing the `Job` interface seems overkill, and will turn it into something much more specific instead of a generic Job handler like it is now.

We could pass in a function instead of a channel
`OnWatcherInit` that would get called when the OpInit event is received.


<details>
  <summary>Something like this</summary>

```diff
diff --git a/integrations/access/accessrequest/app.go b/integrations/access/accessrequest/app.go
index 7bb74a69d9..c575306b17 100644
--- a/integrations/access/accessrequest/app.go
+++ b/integrations/access/accessrequest/app.go
@@ -128,7 +128,7 @@ func (a *App) run(ctx context.Context) error {
 		{Kind: types.KindAccessMonitoringRule},
 	}
 
-	confirmedWatchKindsCh := make(chan []types.WatchKind, 1)
+	acceptedWatchKinds := make([]string, 0, len(watchKinds))
 	job, err := watcherjob.NewJobWithConfirmedWatchKinds(
 		a.apiClient,
 		watcherjob.Config{
@@ -136,7 +136,11 @@ func (a *App) run(ctx context.Context) error {
 			EventFuncTimeout: handlerTimeout,
 		},
 		a.onWatcherEvent,
-		confirmedWatchKindsCh,
+		func(ws types.WatchStatus) {
+			for _, watchKind := range ws.GetKinds() {
+				acceptedWatchKinds = append(acceptedWatchKinds, watchKind.Kind)
+			}
+		},
 	)
 	if err != nil {
 		return trace.Wrap(err)
@@ -149,16 +153,9 @@ func (a *App) run(ctx context.Context) error {
 		return trace.Wrap(err)
 	}
 
-	select {
-	case <-ctx.Done():
-		return trace.Wrap(ctx.Err(), "context is closing")
-	case confirmedKinds := <-confirmedWatchKindsCh:
-		if slices.ContainsFunc(confirmedKinds, func(kind types.WatchKind) bool {
-			return kind.Kind == types.KindAccessMonitoringRule
-		}) {
-			if err := a.initAccessMonitoringRulesCache(ctx); err != nil {
-				return trace.Wrap(err, "initializing Access Monitoring Rule cache")
-			}
+	if slices.Contains(acceptedWatchKinds, types.KindAccessMonitoringRule) {
+		if err := a.initAccessMonitoringRulesCache(ctx); err != nil {
+			return trace.Wrap(err, "initializing Access Monitoring Rule cache")
 		}
 	}
 
diff --git a/integrations/lib/watcherjob/watcherjob.go b/integrations/lib/watcherjob/watcherjob.go
index 760b35a353..3ea92d4841 100644
--- a/integrations/lib/watcherjob/watcherjob.go
+++ b/integrations/lib/watcherjob/watcherjob.go
@@ -43,6 +43,7 @@ const (
 )
 
 type EventFunc func(context.Context, types.Event) error
+type WatchInitFunc func(types.WatchStatus)
 
 type Config struct {
 	Watch            types.Watch
@@ -53,11 +54,11 @@ type Config struct {
 
 type job struct {
 	lib.ServiceJob
-	config                Config
-	eventFunc             EventFunc
-	events                types.Events
-	eventCh               chan *types.Event
-	confirmedWatchKindsCh chan<- []types.WatchKind
+	config          Config
+	eventFunc       EventFunc
+	events          types.Events
+	eventCh         chan *types.Event
+	onWatchInitFunc WatchInitFunc
 }
 
 type eventKey struct {
@@ -71,15 +72,15 @@ func NewJob(client teleport.Client, config Config, fn EventFunc) (lib.ServiceJob
 
 // NewJobWithConfirmedWatchKinds returns a new watcherJob and passes confirmed watch kinds
 // from the initialisation down confirmedWatchKindsCh.
-func NewJobWithConfirmedWatchKinds(events types.Events, config Config, fn EventFunc, confirmedWatchKindsCh chan<- []types.WatchKind) (lib.ServiceJob, error) {
-	return newJobWithEvents(events, config, fn, confirmedWatchKindsCh)
+func NewJobWithConfirmedWatchKinds(events types.Events, config Config, fn EventFunc, watchInitFunc WatchInitFunc) (lib.ServiceJob, error) {
+	return newJobWithEvents(events, config, fn, watchInitFunc)
 }
 
 func NewJobWithEvents(events types.Events, config Config, fn EventFunc) (lib.ServiceJob, error) {
 	return newJobWithEvents(events, config, fn, nil)
 }
 
-func newJobWithEvents(events types.Events, config Config, fn EventFunc, confirmedWatchKindsCh chan<- []types.WatchKind) (job, error) {
+func newJobWithEvents(events types.Events, config Config, fn EventFunc, watchInitFunc WatchInitFunc) (job, error) {
 	if config.MaxConcurrency == 0 {
 		config.MaxConcurrency = DefaultMaxConcurrency
 	}
@@ -94,11 +95,11 @@ func newJobWithEvents(events types.Events, config Config, fn EventFunc, confirme
 		config.FailFast = flag
 	}
 	job := job{
-		events:                events,
-		config:                config,
-		eventFunc:             fn,
-		eventCh:               make(chan *types.Event, config.MaxConcurrency),
-		confirmedWatchKindsCh: confirmedWatchKindsCh,
+		events:          events,
+		config:          config,
+		eventFunc:       fn,
+		eventCh:         make(chan *types.Event, config.MaxConcurrency),
+		onWatchInitFunc: watchInitFunc,
 	}
 	job.ServiceJob = lib.NewServiceJob(func(ctx context.Context) error {
 		process := lib.MustGetProcess(ctx)
@@ -195,19 +196,13 @@ func (job job) watchEvents(ctx context.Context) error {
 
 // waitInit waits for OpInit event be received on a stream.
 func (job job) waitInit(ctx context.Context, watcher types.Watcher, timeout time.Duration) error {
-	var confirmedWatchKinds []types.WatchKind
-	defer func() {
-		if job.confirmedWatchKindsCh != nil {
-			job.confirmedWatchKindsCh <- confirmedWatchKinds
-		}
-	}()
 	select {
 	case event := <-watcher.Events():
 		if event.Type != types.OpInit {
 			return trace.ConnectionProblem(nil, "unexpected event type %q", event.Type)
 		}
 		if watchStatus, ok := event.Resource.(types.WatchStatus); ok {
-			confirmedWatchKinds = watchStatus.GetKinds()
+			job.onWatchInitFunc(watchStatus)
 		}
 		return nil
 	case <-time.After(timeout):

```
</details>

Reply:
That is much better, I also didn't like the channel solution but couldn't think of a better way

Reply:
Can you also update the tests/mock?
The mock is basically duplicating the code, so, I'm not sure what's actually being tested (not related with your PR).

Reply:
Does this look good to you? Needed to add some logic to allow waiting for the callback to run.

Reply:
You should not need it if you add a `j.eventWatcher.WaitReady` (or similar).
This way we keep test and production similar.